### DISCUSSION TITLE
Livewire component vs subcomponent vs trait - use case

### DISCUSSION THREAD

**Discussion by aspyropoulos** [livewire/livewire#3672] [1. Help]
I have the following relationship: Product has many ProductOptions

Which of the three examples is the best to structure my code? Which one will scale better for a big component with many relationships?

Example 1: Both relationships inside same component
```
class Product extends Component
{
    public function addProduct() {}
    public function addProductOptions() {}
}
```

Example 2: Main relationship into my main component. Sub relationship inside a trait.
```
trait withProductOptions
{
    public function addProductOptions() {}
}

class Product extends Component
{
    use withProductOptions
    public function addProduct() {}
}
```

Example 3: Main relationship into my main component. Sub relationship inside sub component.
```
class ProductOptions extends Component
{
    public function addProductOptions() {}
}

class Product extends Component
{
    public function addProduct() {}
}
```

**Comment by joshhanley** [comment]
@aspyropoulos I have used all three in different situations.

It all just depends on you, your code and your application.

Just do what feels best 🙂

I will start with things in one component and then refactor later to either subcomponents (if inter component communication using events is acceptable) or traits (if I need everything to still be in one component).

Hope this helps!

**Reply by aspyropoulos** [reply]
Thanks for the reply @joshhanley. My main concern here is that when using a subcomponents (Ex3) and I need to trigger an action from a main component to a subcomponent and acknowledge it back I need 3 http requests. If you apply that rule to the whole project (wherever is needed) and you have an application used by hundreds or thousands of users then I am worry that I am bombering my server with 3 times more requests where if I was using Ex1 or Ex2 that would only be times 1.

On the other hand for big components Ex1 can lead really fast in a very big class file which I want to avoid.

Lastly in Ex2 although you achieve 1 HTTP request for all the actions and your code is better settled to a trait then again this is not how the traits supposed to be used since the methods there are not necessarily reusable but specific to actions.  

Is there a way I can I achieve 1 request for all my actions and also avoid having a big component by splitting my code in a way other than a trait?

**Reply by joshhanley** [reply]
@aspyropoulos yeah I understand, non of the above is ideal.

Not at the moment, no 😕 it's something we're looking into though, but it won't be implemented any time soon as it requires a fair rework of the request system.
